### ISSUE TITLE
Optimise clean function perf

### ISSUE THREAD

**Issue by crazy4groovy** [lukejacksonn/xyzd#4]
Cool lib!

What about
```js
const clense = (a, b) => {
  b && a.push.apply(a, b)
  return a
}
```
or maybe?
```js
const clense = (a, b) => {
  b && a.push.apply(a, [b])
  return a
}
```

to save some memory via mutating?

**Comment by lukejacksonn** [comment]
Hey, thanks!

I hadn't looked into perf too much yet but this makes sense to me and as good of a place to start as any. So three questions:

- Could we get some numbers on the perf gain?
- Are there any downsides to this approach?
- Is there any other low hanging fruit?

**Comment by Mytrill** [comment]
Maybe at `node[2].reduce(clense, []).map(build(x, y, z))` the `build(x, y, z)` could be cached?
Not sure if that's possible though..

**Comment by crazy4groovy** [comment]
Sounds like a good time to first put tests in place ;)

**Comment by okwolf** [comment]
@crazy4groovy if you are talking about simple unit tests for functional regression testing you're welcome to use my own library [Tead](https://tead.js.org).

For performance testing, I have less knowledge regarding which tools to use.

**Comment by lukejacksonn** [comment]
Hahaa, I am a small way through implementing these.. in Tead like @okwolf suggested! There is only a simple set to start off with. I will push them up after work this evening.

**Comment by okwolf** [comment]
@lukejacksonn welcome to the (somewhat exclusive) Tead club! Let me know how your experience is using it, and what you'd like improved.

**Comment by lukejacksonn** [comment]
@okwolf https://github.com/lukejacksonn/ijk/pull/5

**Comment by lukejacksonn** [comment]
I have added tests to the repo 🔬 which means you can now play in (relative) safety 🎉

I tried running the code you provided @crazy4groovy and it doesn't pass the tests. I thinks the problem is it doesn't supporting the unwrapping of arrays in children arrays; like this:

```js
[ 'main', [
  'Hello World',
  [['li', 1],['li', 2],['li', 3]],
]]
```

Which makes returning a map possible without having to spread it into the parents children array:
```js
const Item = x => ['li', x]
['main', [
  'Hello World',
  [1,2,3].map(Item),
]]
```

That could be fixed with some tweaking but again, it is hard to know the actual benefits until we write some perf tests for it too. This can't be too hard but I have never actually done it so might take some time.

Advice welcomed!

**Comment by lukejacksonn** [comment]
Fixed by #17 thanks @crazy4groovy
